salt/states/boto_cloudtrail.py

```diff
@@ -0,0 +1,353 @@
+# -*- coding: utf-8 -*-
+'''
+Manage CloudTrail Objects
+=================
+
+.. versionadded:: Boron
+
+Create and destroy CloudTrail objects. Be aware that this interacts with Amazon's services,
+and so may incur charges.
+
+This module uses ``boto3``, which can be installed via package, or pip.
+
+This module accepts explicit vpc credentials but can also utilize
+IAM roles assigned to the instance through Instance Profiles. Dynamic
+credentials are then automatically obtained from AWS API and no further
+configuration is necessary. More information available `here
+<http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html>`_.
+
+If IAM roles are not used you need to specify them either in a pillar file or
+in the minion's config file:
+
+.. code-block:: yaml
+
+    vpc.keyid: GKTADJGHEIQSXMKKRBJ08H
+    vpc.key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+
+It's also possible to specify ``key``, ``keyid`` and ``region`` via a profile,
+either passed in as a dict, or as a string to pull from pillars or minion
+config:
+
+.. code-block:: yaml
+
+    myprofile:
+        keyid: GKTADJGHEIQSXMKKRBJ08H
+        key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            region: us-east-1
+
+.. code-block:: yaml
+
+    Ensure trail exists:
+        boto_cloudtrail.present:
+            - Name: mytrail
+            - S3BucketName: mybucket
+            - S3KeyPrefix: prefix
+            - region: us-east-1
+            - keyid: GKTADJGHEIQSXMKKRBJ08H
+            - key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+
+'''
+
+# Import Python Libs
+from __future__ import absolute_import
+import logging
+import os
+import os.path
+import json
+
+# Import Salt Libs
+import salt.utils
+from salt.ext.six import string_types
+
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    '''
+    Only load if boto is available.
+    '''
+    return 'boto_cloudtrail' if 'boto_cloudtrail.exists' in __salt__ else False
+
+
+def present(name, Name,
+           S3BucketName, S3KeyPrefix=None,
+           SnsTopicName=None,
+           IncludeGlobalServiceEvents=True,
+           #IsMultiRegionTrail=None,
+           EnableLogFileValidation=False,
+           CloudWatchLogsLogGroupArn=None,
+           CloudWatchLogsRoleArn=None,
+           KmsKeyId=None,
+           LoggingEnabled=True,
+           Tags=None,
+           region=None, key=None, keyid=None, profile=None):
+    '''
+    Ensure trail exists.
+
+    name
+        The name of the state definition
+
+    Name
```

Comment:
The question here is whether to make the parameter consistent with the salt paradigm, or consistent with the boto3 API (which really just passes through AWS APIs.) Unfortunately AWS uses Name with a capital 'N'. Values returned by API calls like 'describe' will therefore have the capitalized 'Name' as the key. Using 'name' as the argument will make that not be symmetric. Note that the conventions across various APIs are not consistent, some of them start with lower case, but CloudTrail is one that starts with capitals.

I don't have a strong opinion about it. I'll change it if that's the preference. I followed the practice here to be consistent with boto_lambda, where I struggled with this same issue. boto_lambda has a bigger API where some outputs may feed other inputs, so there's more reason to stay consistent between input params and output params. In this module, it won't make a huge difference in usage.

boto_lambda.alias_present is the specific case I'm talking about that I'm trying to be consistent with.